spec/components/pretty_text_spec.rb

```diff
@@ -798,12 +798,26 @@ def extract_urls(text)
       expect(post.excerpt).to eq("hello <a href=\"https://site.com\" rel=\"noopener nofollow ugc\">site</a>")
     end
 
+    it "handles div excerpt at the beginning of a post" do
+      expect(PrettyText.excerpt("<div class='excerpt'>hi</div> test", 100)).to eq('hi')
+      post = Fabricate(:post, raw: "<div class='excerpt'>hi</div> test")
+      expect(post.excerpt).to eq("hi")
+    end
+
     it "handles span excerpt at the beginning of a post" do
       expect(PrettyText.excerpt("<span class='excerpt'>hi</span> test", 100)).to eq('hi')
       post = Fabricate(:post, raw: "<span class='excerpt'>hi</span> test")
       expect(post.excerpt).to eq("hi")
     end
 
+    it "ignores max excerpt length if a div excerpt is specified" do
+      two_hundred = "123456789 " * 20 + "."
+      text = two_hundred + "<div class='excerpt'>#{two_hundred}</div>" + two_hundred
+      expect(PrettyText.excerpt(text, 100)).to eq(two_hundred)
```

Comment:
is it right to ignore? seems a bit odd. I would just truncate at 100

Reply:
It was [implemented](https://github.com/discourse/discourse/blob/d567093756e93ad30c49d520756668a4df902c4c/lib/excerpt_parser.rb#L20) like this when the feature was introduced. So it's just documenting existent implementation.

I had the same doubts, but eventually, I feel that yes, it's a good design decision. 

We always truncate automatic excerpts. But when it comes to custom excerpts it's better to give a user control of this. Probably the user decided to use a custom excerpt just to show that last word or even sentence that was truncated by automatic excerpt extraction.

Reply:
Having read [this](https://meta.discourse.org/t/improving-pinned-topic-excerpts/17640), I think we're using one variable `length` for two different purposes. 

When you create an instance of the `ExcerptParser`, the `length` parameter is the maximum length of the excerpt to be created. 

However, when the parser [detects](https://github.com/discourse/discourse/blob/56717c091077e760875bd2d6c0050d73394e21f9/lib/excerpt_parser.rb#L32) the user added a custom excerpt, it will change the value of that `length` parameter to ensure it's properly picked up even if it's at the end of the post. 

While I like the idea of giving users the ability to provide a custom excerpt anywhere in the post, I do not like that it's not limited in any sort of ways. That's a recipe for abuse 😱 

What I reckon we should do here @AndrewPrigorshnev is when we detect a custom excerpt, we should change the `html` with the content of the custom excerpt instead of changing the `length` of the excerpt. That way we get the best of both worlds: users can add a custom excerpt anywhere in their post using a div or a span, and it's limited according to the site setting.

Reply:
I would be fine if we’d have a modifier for the length of the custom excerpt, like 2x the value of the site setting. But there has to be a limit otherwise it will be abused…

Reply:
Probably it can be abused by somehow making the site parse a lot of very long excerpts.

Reply:
>  if we’d have a modifier for the length of the custom excerpt, like 2x the value of the site setting

What if we use an absolute upper bound instead?
We already have an upper bound for automatic excerpts, it's impossible to set _max excerpt length_ upper than 1000:

<img width="662" alt="Screenshot 2021-05-20 at 22 48 19" src="https://user-images.githubusercontent.com/1274517/119032716-8bc78b80-b9bd-11eb-8f76-2bd508b32bca.png">

We can let custom excerpts bypass _max excerpt length_ but always truncate _all_ excerpts by 1000 chars.

I think we can merge this bug fix and then implement this abuse protection in another PR.

Reply:
That's an excellent idea 👍

Reply:
> We can let custom excerpts bypass max excerpt length but always truncate all excerpts by 1000 chars.

:thinking: not following, are you saying

> We no longer let custom excerpts bypass max excerpt length we truncate at 1000 chars.

Agree with @ZogStriP we need an upper limit otherwise people can sadly grief us and use this escape hatch to break rendering on front page.

Reply:
> 🤔 not following, are you saying
>> We no longer let custom excerpts bypass max excerpt length we truncate at 1000 chars.

Nope, not exactly. An example:

Let's say _topic excerpt maxlength_ set to 220 characters:
<img width="400" alt="Screenshot 2021-05-21 at 13 02 01" src="https://user-images.githubusercontent.com/1274517/119112332-04b5fa00-ba35-11eb-8520-638941900ecc.png">
Then our algorithm for automatic excerpt extraction will take _maximum_ first 220 characters of a post. The rest will be truncated. 

If a user decides to use a custom excerpt (using `<div class="excerpt">` or `<span class="excerpt">`) that excerpt would bypass this restriction. If he wants to add a few more words above the _topic excerpt maxlength_ limit it would work. If he wants to use a custom excerpt of 500 characters for some important pinned topic it would work too.

But to prevent misuse and abuse we will truncate _all_ excerpts by 1000 chars anyway:
* It's already implemented for **automatic excerpts**, users just can't set _topic excerpt maxlength_ to a value bigger than 1000:
  <img width="400" alt="Screenshot 2021-05-21 at 13 15 44" src="https://user-images.githubusercontent.com/1274517/119114140-ec46df00-ba36-11eb-80ab-e0e5eee62a48.png">
* And we will add this protection for **custom excerpts** too. So custom excerpts would bypass _topic excerpt maxlength_ setting but would be truncated on 1000'th character anyway. 

Again I'd like us to add this protection in a separate PR or it would be really hard to follow. We can firstly merge this PR since it's just a bug fix, we just forgot to process `div` in one place and then we can merge another PR with misuse protection.

Reply:
Fine, @SamSaffron  thank you for bringing it up! And @ZogStriP thank you for this nice idea:

> What I reckon we should do here @AndrewPrigorshnev is when we detect a custom excerpt, we should change the html with the content of the custom excerpt instead of changing the length of the excerpt.

 it'll help a lot with implementing this.